### PR TITLE
Fix #2865: Thread should have its own message and posts should only be the replies

### LINKED ISSUE #2865: Thread should have its own message and posts should only be the replies
Threads must be created with a message and today we create the first post with that message and add it to the thread. Instead the thread should have a message and posts should only contain all the replies. This would make it easier to limit the number of posts sent in the response with a limit parameter such as "limitPosts" to get the latest "n" posts in each thread. This will greatly help with reducing the network load and enhance the UX by lazy loading all the posts only when the user clicks on a link/button.

### PR DESCRIPTION
**Pull request by vivekratnavel**
### Describe your changes :
See #2865 

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/UserRepository.java [hunk 1]
```diff
@@ -121,11 +121,13 @@ public User getByEmail(String email, Fields fields) throws IOException, ParseExc
 
   @Override
   public User setFields(User user, Fields fields) throws IOException, ParseException {
-    user.setProfile(fields.contains("profile") ? user.getProfile() : null);
-    user.setTeams(fields.contains("teams") ? getTeams(user) : null);
-    user.setRoles(fields.contains("roles") ? getRoles(user) : null);
-    user.setOwns(fields.contains("owns") ? getOwns(user) : null);
-    user.setFollows(fields.contains("follows") ? getFollows(user) : null);
+    if (fields != null) {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/UserRepository.java:124]: I was under the impression that Fields is never null and we use Fields.EMPTY_FIELDS instead.  

Is there a place where this assumption is not valid?

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java [hunk 2]
```diff
@@ -191,50 +188,60 @@ public ThreadCount getThreadsCount(String link, boolean isResolved) throws IOExc
...
 
-    List<Thread> threads = new ArrayList<>();
-    Set<String> uniqueValues = new HashSet<>();
-    for (String t : threadIds) {
-      // If an entity has multiple relationships (created, mentioned, repliedTo etc.) to the same thread
-      // Don't send duplicated copies of the thread in response
-      if (uniqueValues.add(t)) {
-        threads.add(EntityUtil.validate(t, dao.feedDAO().findById(t), Thread.class));
+  private List<Thread> limitPostsInThreads(List<Thread> threads, int limitPosts) {
+    if (limitPosts == 0) {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:233]: If limit is 0, we return all posts?

Should we do `t.withPosts(null)` when limit = 0? As in return 0 posts per thread

Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:233]: With this patch, both `t.withPosts(null)` and `t.withPosts(0)` will return all the posts in the thread. Do you think we should add support for returning just the thread with 0 posts?

Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:233]: I was possibly misled when I saw that we don't return 0 posts when limit is 0.

From a UX point of view, it seems like we'll always need a non-zero number of posts to show our users. So this logic is good. 

May be add a comment to this line indicating the above, for future reference.

Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:233]: Second thought:
May be also check what is the behavior with pagination and do something similar here.
What happens if we put limit 0 for pagination of GET /tables?

Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:233]: For pagination:

```
      @Parameter(description = "Limit the number tables returned. (1 to 1000000, default = 10)")
          @DefaultValue("10")
          @Min(1)
          @Max(1000000)
          @QueryParam("limit")
```

So all we have to do is document the API accordingly similar to the one above (in FeedResource.java)
Indicate that if limit = 0 it returns everything. -- so this is clear to end users.

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/UserRepository.java (hunk 1 above)
Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/UserRepository.java]: You are right. We don't pass null anywhere for Fields. Removed this null check in the latest patch.

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java (hunk 2 above)
Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java:234]: Sure, added a comment and unit test coverage.

**Review by mithmatt — APPROVED** [review]
Logic looks good.

Need to enhance the feeds threadlist API documentation so that the limit values are clear to the end user.

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java (hunk 2 above)
Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java]: Added the min and max limits in the API for limitPosts and updated the tests. Please take another look. Thank you!

**Review by mithmatt — APPROVED** [review]

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java (hunk 2 above)
Comment (mithmatt) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/FeedRepository.java]: Looks much better with API docs - thanks for incorporating this (especially as I was thinking back and forth through this and providing feedback on the fly)

**Comment by mithmatt** [comment]
Test failure from possible flaky test

```
2022-02-23T22:21:59.7902042Z [INFO] Results:
2022-02-23T22:21:59.7902225Z [INFO] 
2022-02-23T22:21:59.7902398Z [ERROR] Errors: 
2022-02-23T22:21:59.7905259Z [ERROR]   RoleResourceTest.defaultRole:73->createRolesAndSetDefault:124->EntityResourceTest.getEntityByName:1305 » HttpResponse
2022-02-23T22:21:59.7905632Z [INFO] 
2022-02-23T22:21:59.7905869Z [ERROR] Tests run: 891, Failures: 0, Errors: 1, Skipped: 0
```

**Comment by vivekratnavel** [comment]
Test failure is not related to this patch. So, merging it into main. Thanks, @mithmatt for the quick review.

**Pull request merged**